kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/storage/AwsS3DirectoryLister.scala

```diff
@@ -0,0 +1,140 @@
+    continueFrom:     Option[String],
+  )(
+    implicit
+    clock: Clock,
+  ): IO[DirectoryFindResults] =
+    for {
+      listObjsReq <- createListObjectsRequest(bucketAndPrefix, continueFrom)
+      paginator   <- IO(s3Client.listObjectsV2Paginator(listObjsReq))
+      prefixInfo  <- extractPrefixesFromResponse(paginator, exclude, completionConfig)
+      flattened   <- flattenPrefixes(bucketAndPrefix, prefixInfo.partitions, completionConfig, exclude)
+    } yield {
+      prefixInfo match {
+        case CompletedDirectoryFindResults(_)        => CompletedDirectoryFindResults(flattened)
+        case a @ PausedDirectoryFindResults(_, _, _) => a.copy(flattened)
+      }
+    }
+
+  private def flattenPrefixes(
+    bucketAndPrefix:  RemoteS3RootLocation,
+    prefixes:         Set[String],
+    completionConfig: DirectoryFindCompletionConfig,
+    exclude:          Set[String],
+  )(
+    implicit
+    clock: Clock,
+  ): IO[Set[String]] =
+    if (completionConfig.levelsToRecurse == 0) IO.pure(prefixes)
+    else
+      for {
+        ioPrefixes <- IO.pure(prefixes)
+        bap        <- IO.pure(ioPrefixes.map(bucketAndPrefix.fromRoot))
+        recursiveResult <- bap
+          .map(
+            findDirectories(
+              _,
+              completionConfig.copy(levelsToRecurse = completionConfig.levelsToRecurse - 1),
+              exclude,
+              Option.empty,
+            ).map(_.partitions),
+          ).toList.sequence.map {
+            listSetString: List[Set[String]] =>
+              listSetString.toSet.flatten
+          }
+      } yield {
+        recursiveResult.headOption.fold(ioPrefixes)(_ => recursiveResult)
+      }
+
+  private def createListObjectsRequest(
+    bucketAndPrefix: RemoteS3RootLocation,
+    lastFound:       Option[String],
+  ): IO[ListObjectsV2Request] =
+    IO.pure {
+      val builder = ListObjectsV2Request
+        .builder()
+        .maxKeys(1000)
+        .bucket(bucketAndPrefix.bucket)
+        .delimiter("/")
+        .startAfter(lastFound.orNull)
+      bucketAndPrefix.prefix.map(addTrailingSlash).foreach(builder.prefix)
+      builder.build()
+    }
+
+  private def extractPrefixesFromResponse(
+    listObjectsV2Response: ListObjectsV2Iterable,
+    exclude:               Set[String],
+    completionConfig:      DirectoryFindCompletionConfig,
+  ): IO[DirectoryFindResults] =
+    IO {
+      val rtn: mutable.Set[String] = mutable.Set()
+      val resp = listObjectsV2Response
+        .iterator()
+        .asScala
+
+      var partialResultInfo: Option[(String, String)] = none
+
+      while (resp.hasNext && partialResultInfo.isEmpty) {
+        val listResp: ListObjectsV2Response = resp.next()
+        val commonPrefixesFiltered = listResp
+          .commonPrefixes()
+          .asScala
+          .map(_.prefix())
+          .filter(connectorTask.ownsDir)
+          .filterNot(exclude.contains)
+        rtn.addAll(commonPrefixesFiltered)
+
+        partialResultInfo =
```

Comment:
If you use foldLeft as I suggested earlier, you can also accumulate the partialResultInfo without resorting to mutability.